### ISSUE TITLE
SetLoadFields() calculating DATALENGTH of Blob fields even though we didn't implicitly set them as parameters

### ISSUE THREAD

**Issue by mmilince** [microsoft/AL#6836]
**1. Describe the bug**
Function SetLoadFields() will include BLOB fields in SQL query even though we didn't specify those fields as parameters inside SetLoadFields() function.

**2. To Reproduce**

1. Create a new App using AL:GO command
2. Call the SetLoadFields method with the specified field as a parameter

``` AL
page 50100 "Test Page"
{
    Caption = 'Test page';
    PageType = Card;
    ApplicationArea = All;
    UsageCategory = Administration;
    SourceTable = Customer;

    actions
    {
        area(Processing)
        {
            action(Test)
            {
                Caption = 'Test';
                ApplicationArea = All;

                trigger OnAction()
                var
                    Customer: Record Customer;
                begin
                    Customer.SetFilter("No.", '<>%1', Rec."No.");
                    Customer.SetLoadFields("No.");
                    if Customer.FindSet() then
                        repeat
                            Message(Customer.Name);
                        until Customer.Next() = 0;
                end;
            }
        }
    }
}
```

The query for the Customer.FindSet() will be:

```
SELECT "18"."timestamp",
             "18"."No_",
             "18"."Shipping Agent Code",
             "18"."Country_Region Code",
             DATALENGTH("18"."Picture"),
             "18"."Tax Area Code",
             "18"."$systemId",
             "18"."$systemCreatedAt",
             "18"."$systemCreatedBy",
             "18"."$systemModifiedAt",
             "18"."$systemModifiedBy" 
FROM "NP496044".dbo."CRONUS International Ltd_$Customer$437dbf0e-84ff-417a-965d-ed2bb9650972" "18"  WITH(READUNCOMMITTED)  
WHERE ("18"."No_"<>@0) ORDER BY "No_" ASC OPTION(FAST 50)
```

We can notice that besides system fields and those included in calcformulas there is also DATALENGTH("18"."Picture") where the Picture field is of type Blob.
This goes even further because if we create a table extension to the Customer table with a new Blob field like this:

```
tableextension 50100 "PTE Customer" extends Customer
{
    fields
    {
        field(50100; "Test Field"; Blob)
        {
            Caption = 'Test Field';
            DataClassification = CustomerContent;
        }
    }
}
```

Now the SQL query from the Customer.FindSet() function will look like:
```
SELECT "18"."timestamp",
             "18"."No_",
             "18"."Shipping Agent Code",
             "18"."Country_Region Code",
             DATALENGTH("18"."Picture"),
             "18"."Tax Area Code",
             DATALENGTH("18_e5"."Test Field"),
             "18"."$systemId",
             "18"."$systemCreatedAt",
             "18"."$systemCreatedBy",
             "18"."$systemModifiedAt",
             "18"."$systemModifiedBy" 
FROM "NP496044".dbo."CRONUS International Ltd_$Customer$437dbf0e-84ff-417a-965d-ed2bb9650972" "18"  WITH(READUNCOMMITTED)  
JOIN "NP496044".dbo."CRONUS International Ltd_$Customer$ef09dba5-54ac-4a38-ad08-999ed97156c5" "18_e5"  WITH(READUNCOMMITTED)  ON ("18"."No_" = "18_e5"."No_") 
WHERE ("18"."No_"<>@0) ORDER BY "No_" ASC OPTION(FAST 50)
```

Using the JOIN above makes us lose all the performance benefits of SetLoadFields() which was intended to give us in the first place.

**3. Expected behavior**
The ideal expected scenario would be that both of those queries look the same similar to this:
```
SELECT "18"."timestamp",
             "18"."No_",
             "18"."$systemId",
             "18"."$systemCreatedAt",
             "18"."$systemCreatedBy",
             "18"."$systemModifiedAt",
             "18"."$systemModifiedBy" 
FROM "NP496044".dbo."CRONUS International Ltd_$Customer$437dbf0e-84ff-417a-965d-ed2bb9650972" "18"  WITH(READUNCOMMITTED)  
WHERE ("18"."No_"<>@0) ORDER BY "No_" ASC OPTION(FAST 50)
``` 

But as you already mentioned in Remarks of the documentation of SetLoadFields() (https://docs.microsoft.com/en-us/dynamics365/business-central/dev-itpro/developer/methods-auto/record/record-setloadfields-method), calcformulas fields are expected, so realistic expected behavior for both of the above queries would be:

```
SELECT "18"."timestamp",
             "18"."No_",
             "18"."Shipping Agent Code",
             "18"."Country_Region Code",
             "18"."Tax Area Code",
             "18"."$systemId",
             "18"."$systemCreatedAt",
             "18"."$systemCreatedBy",
             "18"."$systemModifiedAt",
             "18"."$systemModifiedBy" 
FROM "NP496044".dbo."CRONUS International Ltd_$Customer$437dbf0e-84ff-417a-965d-ed2bb9650972" "18"  WITH(READUNCOMMITTED)  
WHERE ("18"."No_"<>@0) ORDER BY "No_" ASC OPTION(FAST 50)
```

**4. Versions:**

- AL Language: v8.1.540594
- Visual Studio Code: v1.62.2
- Business Central: W1 19.0 (Platform 19.0.29884.30666 + Application 19.0.29894.30693)
- List of Visual Studio Code extensions that you have installed: -

**Comment by thloke** [comment]
This is a good find. @PhDuck this is something for you I think. Is this expected?

**Comment by dnpb** [comment]
Hopefully this is not how `Blob.HasValue()` works? ;-)

**Comment by PhDuck** [comment]
Thanks for your issue. I talked a bit about it here already:
https://github.com/microsoft/AL/issues/6674#issuecomment-866272925

Unfortunately, the Server and Client were built around the concept that all normal fields would always be loaded, so spread all over you will find assumptions based on this. Because of these assumption we have a bunch of fields that needs to be loaded anyways.

Generally we want to remove some of these assumptions over time, so we cannot give a definitive list that defines which fields will always be included.

The documentation also points it out here:
https://docs.microsoft.com/en-us/dynamics365/business-central/dev-itpro/developer/methods-auto/record/record-setloadfields-method#remarks

**Comment by PhDuck** [comment]
@thloke Please close this issue.

**Comment by dnpb** [comment]
@PhDuck neither of those links refer to BLOB fields, do they?

**Comment by PhDuck** [comment]
@dnpb As I just wrote:
> Generally we want to remove some of these assumptions over time, so **we cannot give a definitive list that defines which fields will always be included.**

The documentation further states:
>Depending on the runtime version, the runtime **may require extra fields** to be selected for loading. Which extra fields to specify depends on the state of the record and table or table extension definition. For example, fields that are filtered upon are always loaded, fields that are referred to in calcformulas in the current table or table extension definition, and so on.

Further it might be complicated if you use Partial Records on pages or reports since that uses metadata to decide which fields are necessary.

**Comment by thloke** [comment]
Closing for now. In the future, I think this is better to be raised on Collaborate or BCIdeas, that way we can get more traction to address these types of issues.

**Comment by PhDuck** [comment]
This was actually fixed for version 21 :)

**Comment by Liniik** [comment]
@PhDuck Really? I still see this behaviour on BC 25.3 in cloud. And it's exactly the reason I have one unnecessary join in the generated query... Should I create new issue for it?

**Comment by jogr-navax** [comment]
Please be advised, that this behaviour has definitely NOT been fixed in BC25 (Plattform 25.0.25866.0 + Anwendung 25.1.25900.0). We had to learn the hard way, that Table Relations on an Extension table that contain a where clause like TableRelation Table.Field where(XXX = field(YYY))) lead to a JOIN. 

In our scenario, we had a code segment for an extended table that read like 

`Rec.SetLoadFields(PrimaryKey);
if Rec.Get(KeyField) then`

This lead to an - in our opinion - unnecessary JOIN statement. I fear that @PhDuck statement merely reflects a change in the official documentation, that does not tell the whole story.

The [Dev IT Pro site](https://learn.microsoft.com/en-us/dynamics365/business-central/dev-itpro/developer/methods-auto/record/record-setloadfields-method) states as followed:

_Depending on the runtime version, the runtime may require extra fields to be selected for loading. Which extra fields to specify depends on the state of the record and table or table extension definition. For example, fields that are filtered upon are always loaded._
